source/_posts/zio-errors.md

```diff
@@ -0,0 +1,177 @@
+ Let's see what our example looks like:
+ ```
+ def forecast(city: String) : F[Either[Error, Double]] =
+ for {
+   location <- EitherT(lookupCity(city))
+   forecast <- EitherT(weather(location))
+   temp <- EitherT.fromEither(getTemperature(forecast).toEither("No temp in forecast"))
+ } yield (temp)
+```
+ 
+Much better! This can be further improved by introducing some syntax, e.g.:
+
+```
+type ErrorOr[E,A] = EitherT[F, E, A]
+
+  implicit class feitherToErrorOr[E, A](r: F[E Either A]) { def toErrorOr: ErrorOr[E, A] = EitherT(r) }
+  implicit class optionToErrorOr[E, A](o: Option[A], e: Error) { def toErrorOr: ErrorOr[E, A] = EitherT.fromEither(r.toEither(e) }
+```
+the code above becomes:
+
+ ```
+ def forecast(city: String) : ErrorOr[Double] =
+ for {
+   location <- lookupCity(city).toErrorOr
+   forecast <-  weather(location).toErrorOr
+   temp <-  getTemperature(forecast).toErrorOr("No temp in forecast"))
+ } yield (temp)
+```
+
+The code above looks much nicer. In fact, I saw this on one of my first Scala projects and I was impressed by its power and purity. 
+
+## ZIO
+
+`ZIO` is an alternative IO implementation. Its approach to errors differs from cats `IO`. 
+Cats has one type parameter - `IO[A]` returns a value of type `A`. It still shortcuts on error by inheriting `MonadError[E]`. 
+`ZIO` chooses to be explicit about the error. It is defined as
+ `ZIO[R,E,A]`.  ZIO is the effectful version of `R => Either[E, A]` - given R, on execution ZIO returns either a value of type 
+ `A` or an error of type `E`. Where `IO[A]` is functorial in `A`, `ZIO[R, E, A]` is functorial in both `E` and `A`. 
+ 
+ If we decided we don't need `R` for now, ZIO provides `type IO[+E, +A] = ZIO[Any, E, A]`. 
+ For our application, given a specific error type `type Error = String`, we would have 
+ `type AppTask[A] = ZIO[Any, Error, A]`. In our ZIO-based application,
+  the business functions would be `lookupCity(city: String]: AppTask[Location]` 
+  and `weather(l: Location]: AppTask[Forecast]` and the code:
+```
+ def forecast(city: String) : AppTask[Double] =
+ for {
+   location <- lookupCity(city)
+   forecast <-  weather(location)
+   temp <-  ZIO.fromEither(getTemperature(forecast).toEither("No temp in forecast"))
+ } yield (temp)
+
+```
+
+# Conclusion
+
+The developers of an application may choose to keep the effect type polymorphic
+throughout the whole application. Today the cats typeclasses are the facto standard, so
+the simplest polimorfic type would be:
+   `F[_] : Sync`. Both `IO` and `ZIO` provide implementation of thes `Sync` type class, and
+   of few more needed for asynchronous I/O.
+   
+The cost to pay for keeping compatibility with any implementation, is the extra steps needing to 
+handle error handling flows - monad transformers, as shown above, or maybe MTL. If, on the other hand,
+the application commits to ZIO, the bifunctor approach to errors results in code
+that is much simpler to write and read.
+
```

Comment:
Nice article @kolov ! I think it sums up a lot of concepts in a digestible way (provided you're not green). One think that I like on technical articles sometimes is if they also provide the imports (because sometimes it's difficult for me to understand where the hell this toEither comes from). But again that's not necessary. Send me the link when you post it!